### DISCUSSION TITLE
Training on more than one objective function

### DISCUSSION THREAD

**Discussion by Juliettengango1** [neuralhydrology/neuralhydrology#240] [Q&A]
Hello everyone, 

Is it possible to train an NH model using more than one loss/obj. function such as NSE, MSE, RMSE. If so, how do you specify such arguments in the configuration file

**Comment by dankovacek** [comment]
Hi @Juliettengango1, 

A similar question occurred to me as well.  If I understand your question, while I see it's possible to track multiple evaluation metrics in the output by listing them in the config file under loss, i.e.:

https://github.com/neuralhydrology/neuralhydrology/blob/922f8b385eb4c736d5cf41d4a427c556af6e61a3/examples/config.yml.example#L44-L51

I'm not sure it's possible to combine two losses in the objective function of the model without creating a new loss object (currently supported losses are listed in the [documentation here](https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.training.html#neuralhydrology.training.get_loss_obj)).   It looks like you can create your own loss object in the `training/loss.py` file.  The loss that the model uses is specified further down in the config:

https://github.com/neuralhydrology/neuralhydrology/blob/922f8b385eb4c736d5cf41d4a427c556af6e61a3/examples/config.yml.example#L153-L154

Based on the example of `MaskedMSELoss` in the `loss.py` script, I would try giving them the same basic structure, i.e. as a starting point:

```[python]
class CombinedNSEKGELoss(BaseLoss):
    """Combined NSE and KGE loss.

    Computes (...) from NSE and KGE metrics.    

    Parameters
    ----------
    cfg : Config
        The run configuration.
    eps: float, optional
        Small constant for numeric stability.
    """

    def __init__(self, cfg: Config, nse_weight: float = 0.5, kge_weight: float = 0.5):
        super().__init__(
            cfg,
            prediction_keys=["y_hat"],
            ground_truth_keys=["y"]
        )
        self.eps = eps


    def _get_loss(self, prediction: Dict[str, torch.Tensor],
                  ground_truth: Dict[str, torch.Tensor],
                  **kwargs):
        mask = ~torch.isnan(y)
        y_hat = prediction["y_hat"][mask]
        y = ground_truth["y"][mask]

        # NSE component
        # ... compute the NSE metric

        # KGE component
        # ... compute the KGE metric
 
        # Combine
        combined_loss = # ...write your combined NSE/KGE loss function
        return combined_loss
```

I'm not sure if there would be other places to register a new loss function, but it's a start anyway.

**Comment by gauchm** [comment]
Exactly what @dankovacek said :)
We currently don't support specifying multiple losses out of the box, but it should be quite easy to write your own combined loss class that reuses the existing loss implementations. Basically, inside the `_get_loss` implementation of your combined loss class, you'd call the `_get_loss` methods of the existing individual losses.
